src/runtime/classbase.cs

```diff
@@ -233,7 +233,75 @@ public static IntPtr tp_str(IntPtr ob)
             }
             try
             {
+                //As per python doc:
+                //The return value must be a string object.  If a class defines __repr__() but not __str__(),
+                //then __repr__() is also used when an “informal” string representation of instances of that
+                //class is required.
+                //In C#, everything provides ToString(), so the check here will be whether the type explicitly
+                //provides ToString() or if it is language provided (i.e. the fully qualified type name as a string)
+
+                //First check which type in the object hierarchy provides ToString()
+                //ToString has two "official" overloads so loop over GetMethods to get the one without parameters
+                var instType = co.inst.GetType();
+                foreach (var method in instType.GetMethods())
+                {
+
+                    //TODO this could probably be done more cleanly with Linq
+                    if (!method.IsPublic) continue; //skip private/protected methods
+                    if (method.Name != "ToString") continue; //only look for ToString
+                    if (method.DeclaringType == typeof(object)) continue; //ignore default from object
+                    if (method.GetParameters().Length != 0) continue; //ignore Formatter overload of ToString
+
+                    //match!  something other than object provides a parameter-less overload of ToString
+                    return Runtime.PyString_FromString(co.inst.ToString());
+                }
+
+                //If the object defines __repr__, call it.
+                System.Reflection.MethodInfo reprMethodInfo = instType.GetMethod("__repr__");
```

Comment:
That's not the same, though. `ToString` is more akin to `__str__`, there is no real equivalent to `__repr__` in .NET.

Reply:
Yeah, this is the `tp_str` implementation. `tp_repr` is below

Reply:
@lostmsu See the comment:
//The return value must be a string object.  If a class defines __repr__() but not __str__(),
//then __repr__() is also used when an “informal” string representation of instances of that
//class is required.

This sentence comes straight from the python manual, I didn't make it up

Reply:
@koubaa the default inheritance hierarchy for C# classes from Python's point of view is Python's `object` <- `System.Object` <- `CustomCSharpClass`.

Since for all .NET classes overriding `ToString` is how you implement `__str__`, I don't see why it should be different for `System.Object`. Hence I think .NET classes should just always call `ToString` for `tp_str`.

Reply:
@lostmsu I  understand your perspective.  If you think about it from a programmer's point of view, in python if you write a class and implement __repr__ and do NOT implement __str__, you will get the __repr__ when you call __str__.  In C# if you write a new class and do NOT implement ToString(), I suggest the same behavior from the point of view of the python interpreter.

In short both languages have a default to_string : 

- python __str__  => '<Foo object at 0xSomething>'
- C# ToString() =>'Namespace.Foo'

neither of them are chosen by the user.  classes which implement __repr__ would see their behavior overwrite the default in python, and should do the same in C# in my opinion

Reply:
@koubaa it is hard to weight these two arguments against each other. They seem like a matter of preference.

But there's an additional one for the status quo: it is much faster, as it does not require reflection at all.

Reply:
@lostmsu sure, its preference.  I'll let others weigh in but I am open to your preference if thats the consensus.

Reply:
@filmor I guess it is just voting. Can you make a decision here?

Reply:
BTW, check out my `__getattr__` implementation here: https://github.com/pythonnet/pythonnet/pull/901/files

Reply:
@lostmsu I took a look.  I agree some kind of consensus between interface vs reflection for dunder methods should be reached and applied throughout the code base.  One argument for reflection is that you can use extension methods to add these.  If I take an existing C# codebase and write some extension methods in a separate assembly, I can provide __repr and __getattr__ to it.  What do you think?

Reply:
@koubaa, extension methods would be hard to find. You'd have to enumerate all types in all assemblies. `GetMethods` won't return them.

In addition to that, what would be the semantics of multiple assemblies defining an extension method with the same signature?

Reply:
@lostmsu ah, didn't know that.

Reply:
@lostmsu I am still not convinced about the interface approach.  It would require a hard build dependency on C# libraries onto PythonNet.  It seems like a tough sell for some of the mature C# libraries out there.

Reply:
@lostmsu I changed this as you requested.  I don't think I can capture the repr function in a Func<string> object since this is a static method.